cmd/hub/mcscontroller/controller.go

```diff
@@ -0,0 +1,169 @@
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceexports/status,verbs=get;update
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceimports,verbs=create;get;list;update;patch;delete
+
+// Reconcile handles an update to a ServiceExport
+// By design, the event that triggered reconciliation is not passed to the reconciler
+// to ensure actions are based on state alone. This approach is referred to as level-based,
+// as opposed to edge-based.
+func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	log := r.Log.WithValues("serviceexport", req.NamespacedName)
+
+	// @todo extract and log the real (i.e., source) object name, namespace and cluster?
+	log.Info("Reconciling ServiceExport")
+
+	se := mcsv1a1.ServiceExport{}
+	if err := r.Client.Get(ctx, req.NamespacedName, &se); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("No ServiceExport found")
+		} else {
+			log.Error(err, "Error fetching ServiceExport")
+		}
+		return ctrl.Result{}, nil
+	}
+
+	result, err := r.reconcile(ctx, &se)
+	if err != nil {
+		return result, err
+	}
+
+	if se.GetDeletionTimestamp() != nil { // service export is marked to be deleted
+		return r.handleDelete(ctx, &se)
+	}
+
+	return result, err
+}
+
+// reconcile the cluster state.
+//
+// Reconciliation logic:
+// 1. list all relevant exports (i.e., having the same name and namespace labels)
+// 2. determine the primary export object
+// 3. iterate on all relevant export objects
+// 3a. if the object is being deleted (non empty deletion timestamp) - ignore
+// 3b. if object conflicts with primary - update export.Status.Condition.Conflict
+//		and delete corresponding ServiceImport.
+// 3c. otherwise, create/update the corrresponding ServiceImport and set the
+//		ServiceExport as owner.
+//
+// Notes:
+// - ensure objects are protected with a finalizer. Since we only receive the
+//	 affected object name into the reconciliation loop, we have no way to determine
+//	 and retrieve affected ServiceExports (i.e., client.List would not return the
+//   deleted object and we can't access its labels). The finalizer is cleared when we
+// 	 receive a notification for change in an object scheduled for deletion.
+// - ServiceImports are owned by the corresponding ServiceExport and are thus
+//	 automatically deleted by k8s when their ServiceExport is deleted.
+//
+func (r *ServiceExportReconciler) reconcile(ctx context.Context, se *mcsv1a1.ServiceExport) (ctrl.Result, error) {
+	lh := mcs.Lighthouse{}
+	name := lh.GetOriginalObjectName(se.ObjectMeta)
+	// @todo: add the NamespacedName of the ServiceExport to be compatible with Reconcile?
+	log := r.Log.WithValues("service", name).WithValues("cluster", lh.GetOriginalObjectCluster(se.ObjectMeta))
+
+	if !controllerutil.ContainsFinalizer(se, serviceExportFinalizer) && se.GetDeletionTimestamp() == nil {
```

Comment:
I think this should be done in the loop iteration all exports and not only for the specific export that triggered the recon.

Reply:
won't we get a notification for them any way and then handle it? Why do it ahead of time? We haven't done any work on a SE before it has a finalizer so if it gets deleted before we have a chance - no harm done. 
I think we can live with that race condition.

Reply:
Moved to Reconcile but still done per object and not trying to process all relevant at once.